### PR TITLE
notify users of card charges

### LINKED ISSUE #1124: Email me whenever you bill my credit card
I just got charged $10, which is fine, but you really need to send me a receipt via email when this happens.  And for extra credit, warn me the day before in a separate email.

This is why I think you need to do this:

1.  Everybody else sends email receipts.  That's really the main reason.  It's a convention now that if you charge somebody's card, especially in an offline background process, you send an email receipt.

2.  It is way easier for me to forward an email to a bookkeeper than it is to log in to a site, save an HTML page and then forward that.  I consider gittip donations a marketing-related business expense.

2.   If you warn me first, it might reduce chargebacks / fraud complaints / failed charges.  You can put a link in the warning email back to my account so that I can immediately cancel my giving.

Keep up the good work!

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/454131-email-me-whenever-you-bill-my-credit-card?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by whit537**
Closes #1124.

#### TODO

- [x] wire up variables for email
- [x] implement text/plain content
- [x] Settings UI
- [ ] review format of received email and tweak as necessary

### CODE REVIEW

**Comment by whit537** [comment]
@chrisdev @citruspi asked for this ahead of PyCon in 10 days.

**Comment by whit537** [comment]
That's towards the bottom of https://github.com/gratipay/inside.gratipay.com/issues/111#issuecomment-83827800.

**Comment by Changaco** [comment]
Ready for review.

**Comment by Changaco** [comment]
Actually, I forgot the settings UI.

**Comment by Changaco** [comment]
UI done, ready for review.

**Comment by whit537** [comment]
UI seems to work as advertised (clicking persists). Testing email ...

**Comment by whit537** [comment]
I think I'm seeing that we're defaulting to sending notifications for failures but not successes. I think we should default to sending notifications for both.

**Comment by whit537** [comment]
No tests for `notify_participants`?

File: gratipay/billing/payday.py [hunk 1]
```diff
@@ -706,6 +707,57 @@ def end(self):
...
+               AND amount > 0
+               AND p.notify_charge > 0
+        """, locals())
+        for e in exchanges:
+            if e.status not in ('failed', 'succeeded'):
+                log('exchange %s has an unexpected status: %s' % (e.id, e.status))
+                continue
+            i = 1 if e.status == 'failed' else 2
+            p = e.participant
+            if p.notify_charge & i == 0:
```
Comment (whit537) [inline gratipay/billing/payday.py]: I find this code difficult to read. You mentioned ([IRC](https://botbot.me/freenode/gratipay/2015-04-06/?msg=35923796&page=1)) wanting to refactor how we're doing notifications. Does this point us in that direction? Are you thinking we'll fold all notifications together into one column that we'll work with via bitmasks? Are you expecting that we'll use bitwise operators with ints? I'd find it easier to comprehend bitwise code if it used `0b00` (Python) and `B'00'` (Postgres). It'd be even easier if we used constants (e.g.,  `NOTIFY_CHARGE_FAILURE`).

**Comment by whit537** [comment]
I get the following error from payday, on a db with two users, one tipping the other $10 (credit card attached and `is_suspicious=False`).

```
pid-47425 thread-140735152444160 (MainThread) Starting a new payday.
pid-47425 thread-140735152444160 (MainThread) Payday started at 2015-04-08 22:21:03.751296+00:00.
pid-47425 thread-140735152444160 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-47425 thread-140735152444160 (MainThread) Prepared the DB.
pid-47425 thread-4368371712 (Thread-1) Holding 1061 cents ($10.00 + $0.61 fee = $10.61) on Balanced for whit537 ... succeeded.
pid-47425 thread-4368371712 (Thread-9) Captured 1000 cents ($9.41 [rounded up from $2.59] + $0.59 fee = $10.00) on Balanced for whit537
pid-47425 thread-140735152444160 (MainThread) Captured 1 card holds.
pid-47425 thread-140735152444160 (MainThread) Canceled 0 card holds.
pid-47425 thread-140735152444160 (MainThread) Updated the balances of 2 participants.
pid-47425 thread-140735152444160 (MainThread) Starting payout loop.
pid-47425 thread-140735152444160 (MainThread) Did payout for 0 participants.
pid-47425 thread-140735152444160 (MainThread) Checked the DB.
pid-47425 thread-140735152444160 (MainThread) Updated payday stats.
pid-47425 thread-140735152444160 (MainThread) Updated receiving amounts.
pid-47425 thread-140735152444160 (MainThread) Traceback (most recent call last):
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/cli.py", line 27, in payday
pid-47425 thread-140735152444160 (MainThread)     Payday.start().run()
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/billing/payday.py", line 144, in run
pid-47425 thread-140735152444160 (MainThread)     self.notify_participants()
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/billing/payday.py", line 751, in notify_participants
pid-47425 thread-140735152444160 (MainThread)     """, locals())
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/__init__.py", line 463, in one
pid-47425 thread-140735152444160 (MainThread)     return cursor.one(sql, parameters, default)
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/cursors.py", line 105, in one
pid-47425 thread-140735152444160 (MainThread)     out = self._some(sql, parameters, lo=0, hi=1)
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/cursors.py", line 127, in _some
pid-47425 thread-140735152444160 (MainThread)     self.execute(sql, parameters)
pid-47425 thread-140735152444160 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/psycopg2/extras.py", line 288, in execute
pid-47425 thread-140735152444160 (MainThread)     return super(NamedTupleCursor, self).execute(query, vars)
pid-47425 thread-140735152444160 (MainThread) ProgrammingError: subquery in FROM must have an alias
pid-47425 thread-140735152444160 (MainThread) LINE 4:                            FROM ( SELECT DISTINCT ON (tippee...
pid-47425 thread-140735152444160 (MainThread)                                         ^
pid-47425 thread-140735152444160 (MainThread) HINT:  For example, FROM (SELECT ...) [AS] foo.
pid-47425 thread-140735152444160 (MainThread) 
```

**Comment by whit537** [comment]
Interesting. Travis shows a failure but it's a js failure, not the failure I'm seeing locally.

**Comment by whit537** [comment]
I force-pushed a cleaner test. It still fails for me in the way noted above. Let's see about Travis.

**Comment by whit537** [comment]
There we go, Travis is showing the failure now. Out of time for now.

File: gratipay/billing/payday.py (hunk 1 above)
Comment (Changaco) [inline gratipay/billing/payday.py]: > Does this point us in that direction?

Yes.

> Are you thinking we'll fold all notifications together into one column that we'll work with via bitmasks?

No, but that may be a better idea than what I had in mind, I'll see if it works.

**Comment by Changaco** [comment]
Ready for more review.

**Comment by Changaco** [comment]
> I think I'm seeing that we're defaulting to sending notifications for failures but not successes. I think we should default to sending notifications for both.

You think the percentage of users who want to receive success notifications is high enough to turn them on by default?

**Comment by kaguillera** [comment]
+ 1 For default notifications of success since most site that you use your credit card for sends you a email when they charge you.

**Comment by whit537** [comment]
Also, my educated guess (haven't verified) is that we have a number of users that tip a small amount so they get charged once initially and then a second time two months later, after they've forgotten about us. I think it's good practice to notify on success by default so that they're notified of that second charge and it doesn't become something they're confused about a year later.

**Comment by Changaco** [comment]
Default changed.

**Comment by whit537** [comment]
Tests pass for me now.

File: gratipay/billing/payday.py (hunk 1 above)
Comment (whit537) [inline gratipay/billing/payday.py]: @Changaco Can we please not use ints for bitmasks? Too clever for me, sorry. :-(

**Comment by whit537** [comment]
Email received:

screen shot 2015-04-13 at 5 03 54 pm

**Comment by whit537** [comment]
Needs some tweaks ...

**Comment by whit537** [comment]
@Changaco We're trying to pull the failure message from `exchanges.note`, but that field is `null` and the error message is in `exchange_routes.error`. Is the error ending up in the wrong field or are we trying to pull it from the wrong location?

**Comment by whit537** [comment]
We should link to the receipt for successful charges.

**Comment by whit537** [comment]
@Changaco Last two things I'm seeing are:

 - [ ] fix error message in failure notification (https://github.com/gratipay/gratipay.com/pull/3301#issuecomment-92518138)
 - [ ] don't use ints for bitmasks (https://github.com/gratipay/gratipay.com/pull/3301#discussion_r28018565)

File: gratipay/billing/payday.py (hunk 1 above)
Comment (rohitpaulk) [inline gratipay/billing/payday.py]: +1 on that. Either not use them at all, or use abstraction so that we don't have to work with bits directly.

**Comment by whit537** [comment]
We'll deal with bitmasks separately, I guess.

**Pull request merged**